# Project-2

# Formal Problem Statement
We want you to predict whether a customer is interested in a caravan insurance policy from other data about the
customer. Information about customers consists of 86 variables and includes product usage data and sociodemographic data derived from zip area codes. The data was supplied based on a real world business problem.
The training set contains over 5000 descriptions of customers, including the information of whether or not they
have a caravan insurance policy. A test set contains 4000 customers of whom target variable is not shared with
you.
Target Variable is V86.
You need to use train data for building the model and then use that model to predict outcome for given test data.
Test dataset does not have a response column; you need to predict those values and submit it in a csv format.
We expect outcomes to be either 0 or 1.

In [1]:
import pandas as pd
import numpy as np


In [2]:
!pip install xgboost

In [2]:
import xgboost as xgb

from sklearn.model_selection import RandomizedSearchCV,train_test_split

from xgboost.sklearn import XGBClassifier 

In [13]:
datafile_train=r"C:\Users\sindhuja\Desktop\python\project\carvan_train.csv"
datafile_test=r"C:\Users\sindhuja\Desktop\python\project\carvan_test.csv"

In [14]:
cd_train=pd.read_csv(datafile_train)
cd_test=pd.read_csv(datafile_test)

In [15]:
cd_train.shape

(5822, 86)

In [16]:
cd_test.shape

(4000, 85)

In [17]:
cd_train.columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21',
       'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31',
       'V32', 'V33', 'V34', 'V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41',
       'V42', 'V43', 'V44', 'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51',
       'V52', 'V53', 'V54', 'V55', 'V56', 'V57', 'V58', 'V59', 'V60', 'V61',
       'V62', 'V63', 'V64', 'V65', 'V66', 'V67', 'V68', 'V69', 'V70', 'V71',
       'V72', 'V73', 'V74', 'V75', 'V76', 'V77', 'V78', 'V79', 'V80', 'V81',
       'V82', 'V83', 'V84', 'V85', 'V86'],
      dtype='object')

In [18]:
cd_test['V86']=np.nan

In [19]:
cd_test.head()

V1  V2  V3  V4  V5  V6  V7  V8  V9  V10  ...  V77  V78  V79  V80  V81  V82  \
0  33   1   4   2   8   0   6   0   3    5  ...    0    0    0    1    0    0   
1   6   1   3   2   2   0   5   0   4    5  ...    0    0    0    1    0    0   
2  39   1   3   3   9   1   4   2   3    5  ...    0    0    0    1    0    0   
3   9   1   2   3   3   2   3   2   4    5  ...    0    0    0    1    0    0   
4  31   1   2   4   7   0   2   0   7    9  ...    0    0    0    1    0    0   

   V83  V84  V85  V86  
0    0    0    0  NaN  
1    0    0    0  NaN  
2    0    0    0  NaN  
3    0    0    0  NaN  
4    0    0    0  NaN  

[5 rows x 86 columns]

In [21]:
cd_train['data']='train'
cd_test['data']='test'

In [22]:
cd_test=cd_test[cd_train.columns] # the columns in the two data frames should be in the same order to enable concatenation

In [23]:
cd_all=pd.concat([cd_train,cd_test],axis=0)

In [24]:
cd_all.head()

V1  V2  V3  V4  V5  V6  V7  V8  V9  V10  ...  V78  V79  V80  V81  V82  V83  \
0  33   1   3   2   8   0   5   1   3    7  ...    0    0    1    0    0    0   
1  37   1   2   2   8   1   4   1   4    6  ...    0    0    1    0    0    0   
2  37   1   2   2   8   0   4   2   4    3  ...    0    0    1    0    0    0   
3   9   1   3   3   3   2   3   2   4    5  ...    0    0    1    0    0    0   
4  40   1   4   2  10   1   4   1   4    7  ...    0    0    1    0    0    0   

   V84  V85  V86   data  
0    0    0  0.0  train  
1    0    0  0.0  train  
2    0    0  0.0  train  
3    0    0  0.0  train  
4    0    0  0.0  train  

[5 rows x 87 columns]

In [25]:
cd_all.tail()

V1  V2  V3  V4  V5  V6  V7  V8  V9  V10  ...  V78  V79  V80  V81  V82  \
3995  33   1   2   4   8   0   7   2   0    5  ...    0    0    1    0    0   
3996  24   1   2   3   5   1   5   1   3    4  ...    0    0    1    0    0   
3997  36   1   2   3   8   1   5   1   3    7  ...    0    0    1    0    0   
3998  33   1   3   3   8   1   4   2   3    7  ...    0    0    0    0    0   
3999   8   1   2   3   2   4   3   0   3    5  ...    0    0    1    0    0   

      V83  V84  V85  V86  data  
3995    0    0    0  NaN  test  
3996    0    0    0  NaN  test  
3997    0    1    0  NaN  test  
3998    0    0    0  NaN  test  
3999    0    0    0  NaN  test  

[5 rows x 87 columns]

In [26]:
cd_all.shape

(9822, 87)

In [27]:
cd_all.dtypes

V1        int64
V2        int64
V3        int64
V4        int64
V5        int64
         ...   
V83       int64
V84       int64
V85       int64
V86     float64
data     object
Length: 87, dtype: object

In [28]:
cd_all.describe()

V1           V2           V3           V4           V5  \
count  9822.000000  9822.000000  9822.000000  9822.000000  9822.000000   
mean     24.253207     1.108735     2.677561     2.996437     5.779067   
std      12.918058     0.412101     0.780701     0.804660     2.874148   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%      10.000000     1.000000     2.000000     2.000000     3.000000   
50%      30.000000     1.000000     3.000000     3.000000     7.000000   
75%      35.000000     1.000000     3.000000     3.000000     8.000000   
max      41.000000    10.000000     6.000000     6.000000    10.000000   

                V6           V7           V8           V9          V10  ...  \
count  9822.000000  9822.000000  9822.000000  9822.000000  9822.000000  ...   
mean      0.700672     4.637650     1.050092     3.262981     6.188964  ...   
std       1.015107     1.721212     1.011156     1.606287     1.896070  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     4.000000     0.000000     2.000000     5.000000  ...   
50%       0.000000     5.000000     1.000000     3.000000     6.000000  ...   
75%       1.000000     6.000000     2.000000     4.000000     7.000000  ...   
max       9.000000     9.000000     5.000000     9.000000     9.000000  ...   

               V77          V78          V79          V80          V81  \
count  9822.000000  9822.000000  9822.000000  9822.000000  9822.000000   
mean      0.004582     0.007941     0.004276     0.574018     0.000916   
std       0.067535     0.088764     0.071224     0.561255     0.030258   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     1.000000     0.000000   
75%       0.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     2.000000     7.000000     1.000000   

               V82         V83          V84          V85          V86  
count  9822.000000  9822.00000  9822.000000  9822.000000  5822.000000  
mean      0.005091     0.03146     0.008450     0.013846     0.059773  
std       0.077996     0.20907     0.092647     0.117728     0.237087  
min       0.000000     0.00000     0.000000     0.000000     0.000000  
25%       0.000000     0.00000     0.000000     0.000000     0.000000  
50%       0.000000     0.00000     0.000000     0.000000     0.000000  
75%       0.000000     0.00000     0.000000     0.000000     0.000000  
max       2.000000     4.00000     2.000000     2.000000     1.000000  

[8 rows x 86 columns]

In [29]:
list(zip(cd_all.columns,cd_all.dtypes,cd_all.nunique()))

[('V1', dtype('int64'), 40),
 ('V2', dtype('int64'), 9),
 ('V3', dtype('int64'), 6),
 ('V4', dtype('int64'), 6),
 ('V5', dtype('int64'), 10),
 ('V6', dtype('int64'), 10),
 ('V7', dtype('int64'), 10),
 ('V8', dtype('int64'), 6),
 ('V9', dtype('int64'), 10),
 ('V10', dtype('int64'), 10),
 ('V11', dtype('int64'), 8),
 ('V12', dtype('int64'), 10),
 ('V13', dtype('int64'), 10),
 ('V14', dtype('int64'), 10),
 ('V15', dtype('int64'), 10),
 ('V16', dtype('int64'), 10),
 ('V17', dtype('int64'), 10),
 ('V18', dtype('int64'), 10),
 ('V19', dtype('int64'), 10),
 ('V20', dtype('int64'), 6),
 ('V21', dtype('int64'), 10),
 ('V22', dtype('int64'), 10),
 ('V23', dtype('int64'), 10),
 ('V24', dtype('int64'), 10),
 ('V25', dtype('int64'), 10),
 ('V26', dtype('int64'), 10),
 ('V27', dtype('int64'), 10),
 ('V28', dtype('int64'), 10),
 ('V29', dtype('int64'), 10),
 ('V30', dtype('int64'), 10),
 ('V31', dtype('int64'), 10),
 ('V32', dtype('int64'), 10),
 ('V33', dtype('int64'), 9),
 ('V34', dtype('int64'), 1

In [30]:
cd_all.columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21',
       'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31',
       'V32', 'V33', 'V34', 'V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41',
       'V42', 'V43', 'V44', 'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51',
       'V52', 'V53', 'V54', 'V55', 'V56', 'V57', 'V58', 'V59', 'V60', 'V61',
       'V62', 'V63', 'V64', 'V65', 'V66', 'V67', 'V68', 'V69', 'V70', 'V71',
       'V72', 'V73', 'V74', 'V75', 'V76', 'V77', 'V78', 'V79', 'V80', 'V81',
       'V82', 'V83', 'V84', 'V85', 'V86', 'data'],
      dtype='object')

In [31]:
cd_all['V1'].value_counts()

33    1401
38     569
8      546
39     542
9      460
3      433
23     376
36     373
35     362
41     355
34     325
24     324
31     318
13     302
11     286
10     271
32     234
37     233
1      218
6      209
12     194
30     190
22     169
2      148
29     139
40     137
25     129
4       90
26      79
27      77
7       72
5       70
20      42
28      41
21      29
18      27
16      25
17      13
15       7
19       7
Name: V1, dtype: int64

In [32]:
cd_all.V2.value_counts()

1     8915
2      821
3       64
7        8
4        4
6        3
5        3
10       2
8        2
Name: V2, dtype: int64

In [33]:
cd_all.V2

0       1
1       1
2       1
3       1
4       1
       ..
3995    1
3996    1
3997    1
3998    1
3999    1
Name: V2, Length: 9822, dtype: int64

In [34]:
cd_all.V4.value_counts() # create dummies since it is age

3    5154
2    2409
4    1777
5     329
1     104
6      49
Name: V4, dtype: int64

In [35]:
cd_all.V5.value_counts() # create dummies

8     2694
3     1513
9     1111
1      959
5      940
7      881
2      827
10     492
6      326
4       79
Name: V5, dtype: int64

In [36]:
cd_all.V6.value_counts()

0    5420
1    2744
2    1213
3     243
4     123
5      30
6      25
7      11
9      10
8       3
Name: V6, dtype: int64

In [37]:
cd_all.V44.value_counts()

0    5903
2    3562
1     341
3      16
Name: V44, dtype: int64

In [38]:

cols = ['V1','V4','V5','V6','V44']
for col in cols:
    k = cd_all[col].value_counts()
    k = k[k>50]
    indexes = k.index[:-1]
    print(indexes)
    for i in indexes:
        print(i)
        cd_all[col+'_'+str(i)] = (cd_all[col]==i).astype(int)

Int64Index([33, 38,  8, 39,  9,  3, 23, 36, 35, 41, 34, 24, 31, 13, 11, 10, 32,
            37,  1,  6, 12, 30, 22,  2, 29, 40, 25,  4, 26, 27,  7],
           dtype='int64')
33
38
8
39
9
3
23
36
35
41
34
24
31
13
11
10
32
37
1
6
12
30
22
2
29
40
25
4
26
27
7
Int64Index([3, 2, 4, 5], dtype='int64')
3
2
4
5
Int64Index([8, 3, 9, 1, 5, 7, 2, 10, 6], dtype='int64')
8
3
9
1
5
7
2
10
6
Int64Index([0, 1, 2, 3], dtype='int64')
0
1
2
3
Int64Index([0, 2], dtype='int64')
0
2


In [39]:
cd_all.head()

V1  V2  V3  V4  V5  V6  V7  V8  V9  V10  ...  V5_7  V5_2  V5_10  V5_6  \
0  33   1   3   2   8   0   5   1   3    7  ...     0     0      0     0   
1  37   1   2   2   8   1   4   1   4    6  ...     0     0      0     0   
2  37   1   2   2   8   0   4   2   4    3  ...     0     0      0     0   
3   9   1   3   3   3   2   3   2   4    5  ...     0     0      0     0   
4  40   1   4   2  10   1   4   1   4    7  ...     0     0      1     0   

   V6_0  V6_1  V6_2  V6_3  V44_0  V44_2  
0     1     0     0     0      1      0  
1     0     1     0     0      0      1  
2     1     0     0     0      0      1  
3     0     0     1     0      1      0  
4     0     1     0     0      1      0  

[5 rows x 137 columns]

In [40]:
cd_all.columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       ...
       'V5_7', 'V5_2', 'V5_10', 'V5_6', 'V6_0', 'V6_1', 'V6_2', 'V6_3',
       'V44_0', 'V44_2'],
      dtype='object', length=137)

### data prep is done   and        following things
1.split the train and test
2.split train into two( use stratified sampling method)
3.build a model on train data


In [41]:
cd_all.shape

(9822, 137)

In [42]:
cd_train.isnull().sum().sum(),cd_test.isnull().sum().sum()

(0, 4000)

In [43]:
cd_all.isnull().sum().sum() # v86 has NaN

4000

In [44]:
cd_test.shape

(4000, 87)

In [45]:
pd.set_option('display.max_rows',500)
cd_all.columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       ...
       'V5_7', 'V5_2', 'V5_10', 'V5_6', 'V6_0', 'V6_1', 'V6_2', 'V6_3',
       'V44_0', 'V44_2'],
      dtype='object', length=137)

In [46]:
cd_all.columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       ...
       'V5_7', 'V5_2', 'V5_10', 'V5_6', 'V6_0', 'V6_1', 'V6_2', 'V6_3',
       'V44_0', 'V44_2'],
      dtype='object', length=137)

# copy the data

In [47]:
cd_copy=cd_all.copy()

In [48]:
cd_copy.head()

V1  V2  V3  V4  V5  V6  V7  V8  V9  V10  ...  V5_7  V5_2  V5_10  V5_6  \
0  33   1   3   2   8   0   5   1   3    7  ...     0     0      0     0   
1  37   1   2   2   8   1   4   1   4    6  ...     0     0      0     0   
2  37   1   2   2   8   0   4   2   4    3  ...     0     0      0     0   
3   9   1   3   3   3   2   3   2   4    5  ...     0     0      0     0   
4  40   1   4   2  10   1   4   1   4    7  ...     0     0      1     0   

   V6_0  V6_1  V6_2  V6_3  V44_0  V44_2  
0     1     0     0     0      1      0  
1     0     1     0     0      0      1  
2     1     0     0     0      0      1  
3     0     0     1     0      1      0  
4     0     1     0     0      1      0  

[5 rows x 137 columns]

## cols = ['V1','V4','V5','V6','V44'] drop this col since we create dummmies for thi

In [49]:
cd_copy.drop( ['V1','V4','V5','V6','V44'],axis=1,inplace=True)

In [50]:
cd_copy.head()

V2  V3  V7  V8  V9  V10  V11  V12  V13  V14  ...  V5_7  V5_2  V5_10  V5_6  \
0   1   3   5   1   3    7    0    2    1    2  ...     0     0      0     0   
1   1   2   4   1   4    6    2    2    0    4  ...     0     0      0     0   
2   1   2   4   2   4    3    2    4    4    4  ...     0     0      0     0   
3   1   3   3   2   4    5    2    2    2    3  ...     0     0      0     0   
4   1   4   4   1   4    7    1    2    2    4  ...     0     0      1     0   

   V6_0  V6_1  V6_2  V6_3  V44_0  V44_2  
0     1     0     0     0      1      0  
1     0     1     0     0      0      1  
2     1     0     0     0      0      1  
3     0     0     1     0      1      0  
4     0     1     0     0      1      0  

[5 rows x 132 columns]

#### split

In [51]:
cd_train=cd_copy[cd_copy['data']=='train']
del cd_train['data']

In [52]:
cd_train.head()

V2  V3  V7  V8  V9  V10  V11  V12  V13  V14  ...  V5_7  V5_2  V5_10  V5_6  \
0   1   3   5   1   3    7    0    2    1    2  ...     0     0      0     0   
1   1   2   4   1   4    6    2    2    0    4  ...     0     0      0     0   
2   1   2   4   2   4    3    2    4    4    4  ...     0     0      0     0   
3   1   3   3   2   4    5    2    2    2    3  ...     0     0      0     0   
4   1   4   4   1   4    7    1    2    2    4  ...     0     0      1     0   

   V6_0  V6_1  V6_2  V6_3  V44_0  V44_2  
0     1     0     0     0      1      0  
1     0     1     0     0      0      1  
2     1     0     0     0      0      1  
3     0     0     1     0      1      0  
4     0     1     0     0      1      0  

[5 rows x 131 columns]

In [53]:
cd_train.shape

(5822, 131)

In [54]:
cd_test=cd_copy[cd_copy['data']=='test']
del cd_test['data']
del cd_test['V86']


In [95]:
x_test=cd_test.copy()

In [96]:
x_test.head(2)

V2  V3  V7  V8  V9  V10  V11  V12  V13  V14  ...  V5_7  V5_2  V5_10  V5_6  \
0   1   4   6   0   3    5    0    4    1    1  ...     0     0      0     0   
1   1   3   5   0   4    5    2    2    1    4  ...     0     1      0     0   

   V6_0  V6_1  V6_2  V6_3  V44_0  V44_2  
0     1     0     0     0      0      0  
1     1     0     0     0      0      1  

[2 rows x 130 columns]

In [55]:
cd_test.head()

V2  V3  V7  V8  V9  V10  V11  V12  V13  V14  ...  V5_7  V5_2  V5_10  V5_6  \
0   1   4   6   0   3    5    0    4    1    1  ...     0     0      0     0   
1   1   3   5   0   4    5    2    2    1    4  ...     0     1      0     0   
2   1   3   4   2   3    5    2    3    2    3  ...     0     0      0     0   
3   1   2   3   2   4    5    4    1    2    4  ...     0     0      0     0   
4   1   2   2   0   7    9    0    0    0    6  ...     1     0      0     0   

   V6_0  V6_1  V6_2  V6_3  V44_0  V44_2  
0     1     0     0     0      0      0  
1     1     0     0     0      0      1  
2     0     1     0     0      0      1  
3     0     0     1     0      0      1  
4     1     0     0     0      0      1  

[5 rows x 130 columns]

In [56]:
cd_test.shape

(4000, 130)

# XGB

In [57]:
x_train=cd_train.drop('V86',axis=1)
y_train=cd_train['V86']

In [58]:
y_train.shape

(5822,)

In [59]:
param_dist = {
              "max_depth": [2,3,4,5,6],
              "learning_rate":[0.01,0.05,0.1,0.3,0.5],
                "min_child_weight":[4,5,6],
              "subsample":[i/10.0 for i in range(6,10)],
                "colsample_bytree":[i/10.0 for i in range(6,10)],
               "reg_alpha":[1e-5, 1e-2, 0.1, 1, 100],
              "gamma":[i/10.0 for i in range(0,5)],
            "n_estimators":[100,500,700,1000],
            'scale_pos_weight':[2,3,4,5,6,7,8,9]
              }

In [60]:
clf=XGBClassifier(objective='binary:logistic')

In [61]:
from sklearn.metrics import make_scorer,fbeta_score

fBeta=make_scorer(fbeta_score,beta=2)

In [ ]:
from sklearn.metrics import make_scorer,fbeta_score

fBeta=make_scorer(fbeta_score,beta=2)

In [62]:
n_iter=25

random_search=RandomizedSearchCV(clf,n_jobs=-1,verbose=3,cv=10,n_iter=n_iter,scoring=fBeta,
                                 param_distributions=param_dist)

In [63]:
random_search.fit(x_train,y_train)

Fitting 10 folds for each of 25 candidates, totalling 250 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  24 tasks      | elapsed:  3.5min
[Parallel(n_jobs=-1)]: Done 120 tasks      | elapsed: 15.0min
[Parallel(n_jobs=-1)]: Done 250 out of 250 | elapsed: 26.8min finished


RandomizedSearchCV(cv=10, error_score=nan,
                   estimator=XGBClassifier(base_score=None, booster=None,
                                           colsample_bylevel=None,
                                           colsample_bynode=None,
                                           colsample_bytree=None, gamma=None,
                                           gpu_id=None, importance_type='gain',
                                           interaction_constraints=None,
                                           learning_rate=None,
                                           max_delta_step=None, max_depth=None,
                                           min_child_weight=None, missing=nan,
                                           monotone_constraints=None,...
                                        'learning_rate': [0.01, 0.05, 0.1, 0.3,
                                                          0.5],
                                        'max_depth': [2, 3, 4, 5, 6],
          

In [81]:
def report(results,n_top=3):
    for i in range(1,n_top+1):
        candidates = np.flatnonzero(results['rank_test_score']==i)
        for candidate in candidates:
            print("Model with rank: {0}".format(i))
            print("Mean Validation Score: {0:.8f} (std:{1:.3f})".format(
                results['mean_test_score'][candidate],
                results['std_test_score'][candidate]))
            print("Parameters: {0}".format(results['params'][candidate]))
            print("")

In [85]:
report(random_search.cv_results_,5)

Model with rank: 1
Mean Validation Score: 0.33747322 (std:0.047)
Parameters: {'subsample': 0.7, 'scale_pos_weight': 8, 'reg_alpha': 100, 'n_estimators': 1000, 'min_child_weight': 6, 'max_depth': 5, 'learning_rate': 0.1, 'gamma': 0.2, 'colsample_bytree': 0.6}

Model with rank: 2
Mean Validation Score: 0.33093873 (std:0.063)
Parameters: {'subsample': 0.7, 'scale_pos_weight': 8, 'reg_alpha': 100, 'n_estimators': 100, 'min_child_weight': 6, 'max_depth': 4, 'learning_rate': 0.5, 'gamma': 0.3, 'colsample_bytree': 0.9}

Model with rank: 3
Mean Validation Score: 0.32924604 (std:0.058)
Parameters: {'subsample': 0.9, 'scale_pos_weight': 8, 'reg_alpha': 100, 'n_estimators': 1000, 'min_child_weight': 5, 'max_depth': 5, 'learning_rate': 0.5, 'gamma': 0.2, 'colsample_bytree': 0.7}

Model with rank: 4
Mean Validation Score: 0.32082880 (std:0.042)
Parameters: {'subsample': 0.9, 'scale_pos_weight': 8, 'reg_alpha': 0.1, 'n_estimators': 100, 'min_child_weight': 5, 'max_depth': 5, 'learning_rate': 0.01, '

In [86]:
# here we choosing rank 4 since it has low std and good validation score comparatively

Model with rank: 4
Mean Validation Score: 0.32082880 (std:0.042)
Parameters: {'subsample': 0.9, 'scale_pos_weight': 8, 'reg_alpha': 0.1, 'n_estimators': 100, 'min_child_weight': 5, 'max_depth': 5, 'learning_rate': 0.01, 'gamma': 0.0, 'colsample_bytree': 0.7}

In [89]:
xgb_best=XGBClassifier(subsample=0.9,scale_pos_weight=8,reg_alpha=0.1,n_estimators=100,min_child_weight=5,
                       max_depth=5,learning_rate=0.01,gamma=0.0,colsample_bytree=0.7,n_jobs=-1)

In [90]:
xgb_best.fit(x_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.7, gamma=0.0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.01, max_delta_step=0, max_depth=5,
              min_child_weight=5, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=-1, num_parallel_tree=1,
              objective='binary:logistic', random_state=0, reg_alpha=0.1,
              reg_lambda=1, scale_pos_weight=8, subsample=0.9,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [93]:
train_score=xgb_best.predict_proba(x_train)[:,1]

In [94]:
train_score

array([0.40580195, 0.3291007 , 0.33727175, ..., 0.45892572, 0.3814503 ,
       0.30467078], dtype=float32)

In [99]:
test_score=xgb_best.predict_proba(cd_test)[:,1]

In [101]:
test_score


array([0.27727517, 0.6439308 , 0.50589967, ..., 0.5205565 , 0.43315312,
       0.33708316], dtype=float32)

## we need hard class

In [102]:
cutoffs=np.linspace(0.01,0.99,99)
cutoffs

array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 , 0.11,
       0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 , 0.21, 0.22,
       0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 , 0.31, 0.32, 0.33,
       0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 , 0.41, 0.42, 0.43, 0.44,
       0.45, 0.46, 0.47, 0.48, 0.49, 0.5 , 0.51, 0.52, 0.53, 0.54, 0.55,
       0.56, 0.57, 0.58, 0.59, 0.6 , 0.61, 0.62, 0.63, 0.64, 0.65, 0.66,
       0.67, 0.68, 0.69, 0.7 , 0.71, 0.72, 0.73, 0.74, 0.75, 0.76, 0.77,
       0.78, 0.79, 0.8 , 0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87, 0.88,
       0.89, 0.9 , 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99])

In [104]:
train_score=xgb_best.predict_proba(x_train)[:,1]
real=y_train


In [105]:
from sklearn.metrics import fbeta_score

In [108]:
fbeta_all=[]
Beta=2
KS_all=[]


for cutoff in cutoffs:
     
        predicated=(train_score>cutoff).astype(int)
        
        TP=((predicated==1) & (real==1)).sum()
        TN=((predicated==0) & (real==0)).sum()
        FP=((predicated==1) & (real==0)).sum()
        FN=((predicated==0) & (real==1)).sum()
        
        P=TP+FN
        N=TN+FP
        
        KS=(TP/P)-(FP/N)
        KS_all.append(KS)
        
        Pr=TP/(TP+FP)
        Recall=TP/(TP+FN)
        
#         F2_numorator=((1+(Beta**2))*Pr )*Recall
#         F2_denomenator=((Beta**2)*Pr)+Recall
        
        F2_Score= (((1+(Beta**2))*Pr )*Recall)/(((Beta**2)*Pr)+Recall)
        
        fbeta_all.append(F2_Score)
      
    

C:\Users\sindhuja\anaconda3\lib\site-packages\ipykernel_launcher.py:21: RuntimeWarning: invalid value encountered in longlong_scalars


In [109]:
list(zip(cutoffs,fbeta_all))

[(0.01, 0.2411976711948988),
 (0.02, 0.2411976711948988),
 (0.03, 0.2411976711948988),
 (0.04, 0.2411976711948988),
 (0.05, 0.2411976711948988),
 (0.060000000000000005, 0.2411976711948988),
 (0.06999999999999999, 0.2411976711948988),
 (0.08, 0.2411976711948988),
 (0.09, 0.2411976711948988),
 (0.09999999999999999, 0.2411976711948988),
 (0.11, 0.2411976711948988),
 (0.12, 0.2411976711948988),
 (0.13, 0.2411976711948988),
 (0.14, 0.2411976711948988),
 (0.15000000000000002, 0.2411976711948988),
 (0.16, 0.2411976711948988),
 (0.17, 0.2411976711948988),
 (0.18000000000000002, 0.2411976711948988),
 (0.19, 0.2411976711948988),
 (0.2, 0.24133148404993066),
 (0.21000000000000002, 0.2448290417897847),
 (0.22, 0.255356618726152),
 (0.23, 0.274318145987703),
 (0.24000000000000002, 0.2892768079800499),
 (0.25, 0.3003625064733299),
 (0.26, 0.308839190628328),
 (0.27, 0.3180568285976169),
 (0.28, 0.33091048201989287),
 (0.29000000000000004, 0.3438624550179928),
 (0.3, 0.35953177257525076),
 (0.31, 0.3

In [110]:
mycutoff=cutoffs[fbeta_all==max(fbeta_all)][0]
mycutoff

0.46

In [111]:
test_score=xgb_best.predict_proba(cd_test)[:,1]
test_score

array([0.27727517, 0.6439308 , 0.50589967, ..., 0.5205565 , 0.43315312,
       0.33708316], dtype=float32)

In [114]:
predictions=(test_score>mycutoff).astype(int)

In [115]:
pd.Series(predictions).value_counts()

0    3335
1     665
dtype: int64

In [116]:
submissions=pd.DataFrame({'V86':predictions})

In [117]:
submissions.to_csv('Surendran_R_project2.csv',index=False)

In [118]:
r=pd.read_csv('Surendran_R_project2.csv')

In [119]:
r.head()

V86
0    0
1    1
2    1
3    0
4    0